### DISCUSSION TITLE
Nuget error when installing HostFx Sample

### DISCUSSION THREAD

**Discussion by johdav999** [NuGet/Home#12896] [General]
I am trying to build the HostFx sample (https://github.com/dotnet/samples/tree/main/core/hosting)
but am getting the error below.
When opening the example I'm upgrading the Platform Toolset to v143 and the Windows SDK to 10.0.22621.1778

I have tried to clean the project, delete bin and obj folder and rebuild project with no result. Anyone know the reason and/or a possible solution?
____________________

Assembly loaded during TaskRun: Microsoft.NuGet.Build.Tasks, Version=17.2.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35 (location: C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Microsoft\NuGet\17.0\Microsoft.NuGet.Build.Tasks.dll, MVID: 7aa05f9d-a90d-4b04-8040-2d758974bac6, AppDomain: [Default])
1> Using "ResolveNuGetPackageAssets" task from assembly "C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Microsoft\NuGet\17.0\Microsoft.NuGet.Build.Tasks.dll".
1> Task "ResolveNuGetPackageAssets"
1> Task Parameter:RuntimeIdentifier=win-x86
1> Task Parameter:IncludeFrameworkReferences=True
1> Task Parameter:
1> ContentPreprocessorValues=
1> rootnamespace
1> Value=NativeHost
1> assemblyname
1> Value=NativeHost.vs
1> fullpath
1> Value=C:\UE\hosting\src\NativeHost
1> outputfilename
1> Value=NativeHost.vs.exe
1> filename
1> Value=NativeHost.vs.vcxproj
1> Task Parameter:ContentPreprocessorOutputDirectory=Debug\NuGet
1> Task Parameter:ProjectLockFile=obj\project.assets.json
1> Task Parameter:ProjectLanguage=C++
1> Assembly loaded during TaskRun (Microsoft.NuGet.Build.Tasks.ResolveNuGetPackageAssets): Newtonsoft.Json, Version=13.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed (location: C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Microsoft\NuGet\17.0\Newtonsoft.Json.dll, MVID: f7499511-9334-484f-b6db-6825be2eac0d, AppDomain: [Default])
1> Assembly loaded during TaskRun (Microsoft.NuGet.Build.Tasks.ResolveNuGetPackageAssets): System.Numerics, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089 (location: C:\WINDOWS\Microsoft.Net\assembly\GAC_MSIL\System.Numerics\v4.0_4.0.0.0__b77a5c561934e089\System.Numerics.dll, MVID: 0cfb80c7-63d2-4fc1-af85-4f9d342b6d09, AppDomain: [Default])
1> C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Microsoft\NuGet\17.0\Microsoft.NuGet.targets(198,5): error : Sequence contains no elements
1> C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Microsoft\NuGet\17.0\Microsoft.NuGet.targets(198,5): error : at System.Linq.Enumerable.First[TSource](IEnumerable`1 source)
1> C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Microsoft\NuGet\17.0\Microsoft.NuGet.targets(198,5): error : at Microsoft.NuGet.Build.Tasks.ResolveNuGetPackageAssets.GiveErrorForMissingFramework()
1> C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Microsoft\NuGet\17.0\Microsoft.NuGet.targets(198,5): error : at Microsoft.NuGet.Build.Tasks.ResolveNuGetPackageAssets.GetTargetOrAttemptFallback(JObject lockFile, Boolean needsRuntimeIdentifier)
1> C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Microsoft\NuGet\17.0\Microsoft.NuGet.targets(198,5): error : at Microsoft.NuGet.Build.Tasks.ResolveNuGetPackageAssets.GetReferences(JObject lockFile)
1> C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Microsoft\NuGet\17.0\Microsoft.NuGet.targets(198,5): error : at Microsoft.NuGet.Build.Tasks.ResolveNuGetPackageAssets.ExecuteCore()
1> C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Microsoft\NuGet\17.0\Microsoft.NuGet.targets(198,5): error : at Microsoft.NuGet.Build.Tasks.ResolveNuGetPackageAssets.Execute()
1> Done executing task "ResolveNuGetPackageAssets" -- FAILED.

**Comment by zivkan** [comment]
@johdav999 

I think the problem is that the sample was not updated as tooling changed.

Historically, C++ projects (`.vcxproj`) didn't support NuGet's PackageReference, hence the C++ project wouldn't run `Microsoft.NuGet.targets`. However at some point during VS 2022, C++/CLI projects started supporting PackageReference. It would appear that the C++ team's implementation causes all vcxproj projects to load NuGet's assets file, even for native C++ projects.

NuGet has [a bug/design choice](https://github.com/NuGet/Home/issues/4463) where PackageReference restore always names one of restore's output files as `$(MSBuildProjectExtensionsPath)\project.assets.json`, which means that we do not support two different projects using the same `$(MSBuildProjectExtensionsPath)` directory.

The sample has both a `csproj` and ` vcxproj` in the same directory, so the combination of the above two issues means that this sample no longer works, but would have worked in the past.

You can work around the issue by creating a `Directory.Build.props` and set the `MSBuildProjectExtensionsPath` property to a different value for each project (either use two properties, each with a condition checking `MSBuildProjectName`, or use `$(MSBuildProjectName)` within `MSBuildProjectExtensionsPath`'s value)

The NuGet team prioritize work based on a number of signals, one of which is [the number of upvotes](https://github.com/NuGet/Home/blob/dev/Issue-Triage-Policy.md#up-voting-a-feature-requestbug) (👍 reactions), so upvoting #4463 might get it prioritized more quickly.

If you believe that vcxproj projects shouldn't use Microsoft.NuGet.targets on project types that don't support PackageReference, then please search for an issue on https://developercommunity.visualstudio.com and upvote it, or create a new issue using Visual Studio's Report a Problem feature.

If you'd like the error message to be more meaningful, please report the issue at https://github.com/dotnet/NuGet.BuildTasks. I know it's confusing, but another team own that repo (which contains `Microsoft.NuGet.targets`), which is responsible for reading the assets file that NuGet restore creates. GitHub only provide us tooling to transfer issues within the same org, so I can transfer to other NuGet/* repos, but not dotnet/* repos.
 
If you'd like the sample to be updated to work with current tooling, please create an issue in https://github.com/dotnet/samples/issues.

**Reply by johdav999** [reply]
Many thanks Zivkan,
I tested your solution to create a Directory.Build.props file and set the MSBuildProjectExtensionsPath property to different values and it worked fine.
For future reference if someone experiences the same issue. Add a Directory.Build.props file to your NativeHost.vs project and in that file add the following:
```
<Project>
  <PropertyGroup Condition="'$(MSBuildProjectName)' == 'App'">
    <MSBuildProjectExtensionsPath>$(MSBuildThisFileDirectory)MyCSharpProject\</MSBuildProjectExtensionsPath>
  </PropertyGroup>
  <PropertyGroup Condition="'$(MSBuildProjectName)' == 'DotNetLib'">
    <MSBuildProjectExtensionsPath>$(MSBuildThisFileDirectory)MyCSharpProject\</MSBuildProjectExtensionsPath>
  </PropertyGroup>

  <PropertyGroup Condition="'$(MSBuildProjectName)' == 'NativeHost.vs'">
    <MSBuildProjectExtensionsPath>$(MSBuildThisFileDirectory)MyCppProject\</MSBuildProjectExtensionsPath>
  </PropertyGroup>
</Project>
```



I also had to add a reference in Project Properties-Linker-General-Additional Library Directories to the folder where the nethost.lib is located. In my case:

\.nuget\packages\microsoft.netcore.app.host.win-x64\7.0.11\runtimes\win-x64\native;%(AdditionalLibraryDirectories)

Then it works to build the sample.
